hydradx/model/amm/arbitrage_agent_general.py

```diff
@@ -0,0 +1,549 @@
+                ex_2_sell_price = ex_2.sell_spot(tkn_pair_2[1], tkn_pair_2[0]) * (1 - arb_cfg['buffer'])
+                if ex_2_sell_price and ex_2_sell_price > ex_1_buy_price:  # buy from ex1, sell to ex2
+                    arb_opps.append(((ex_2_sell_price - ex_1_buy_price) / ex_1_buy_price, i))
+
+    arb_opps.sort(key=lambda x: x[0], reverse=True)
+    return arb_opps
+
+
+def get_arb_swaps(
+        exchanges: dict[str: AMM],
+        config: list[dict],
+        max_liquidity: dict[str: dict[str: float]] = None,
+        max_iters: int = 10,
+        precision: float = 1e-10
+):
+    """
+    return a list of swaps that can be executed to take advantage of the best arbitrage opportunities
+    """
+    if max_liquidity is None:
+        max_liquidity = {ex_name: {tkn: float('inf') for tkn in ex.asset_list} for ex_name, ex in exchanges.items()}
+
+    arb_opps = get_arb_opps(exchanges, config, max_liquidity)
+    all_swaps = []
+    test_exchanges = {ex_name: ex.copy() for ex_name, ex in exchanges.items()}
+    test_agents = {ex_name: Agent(holdings=max_liquidity[ex_name].copy()) for ex_name in exchanges}
+    while arb_opps:
+        arb_cfg = config[arb_opps[0][1]]
+        swap = process_next_swap(
+            exchanges=test_exchanges,
+            swap_config=arb_cfg,
+            agents=test_agents,
+            max_liquidity=max_liquidity,
+            max_iters=max_iters,
+            precision=precision
+        )
+        for exchange in test_exchanges.values():
+            if exchange.fail:
+                raise AssertionError('max liquidity exceeded?')
+        if swap:
+            all_swaps.append(swap)
+        else:
+            break
+        new_arb_opps = get_arb_opps(test_exchanges, config, max_liquidity)
+        if arb_opps and new_arb_opps and arb_opps[0][0] == new_arb_opps[0][0]:
+            break
+        arb_opps = new_arb_opps
+
+    return all_swaps
+
+
+def process_next_swap(
+        exchanges: dict[str: AMM],
+        swap_config: dict,
+        agents: dict[str: Agent],
+        max_liquidity: dict[str: [dict[str: float]]],
+        precision: float = 1e-10,
+        max_iters: int = 10
+):
+    """
+    Calculate what the appropriate size of the swap should be, and execute it on the virtual exchanges
+    and virtual agents to keep track of the results.
+    """
+    buffer = swap_config['buffer']
+    exchange_names = list(swap_config['exchanges'].keys())
+    tkn_pairs = swap_config['exchanges']
+    slippage_tolerance = {ex_name: buffer / 2 for ex_name in exchange_names}
+
+    # make sure the assets are specified in the correct order
+    for ex_name in swap_config['exchanges']:
+        if isinstance(exchanges[ex_name], CentralizedMarket):
+            if tkn_pairs[ex_name] not in exchanges[ex_name].order_book:
+                raise ValueError(f'Asset pair {tkn_pairs[ex_name]} not found in order book. Are they ordered wrong?')
+
+    # all denominated in tkn_pair[1]
+    buy_price = {
```

Comment:
It is a little weird, but I guess I didn't have a better idea at the time, and I didn't put a lot of thought into it because it's only used in this narrow scope. I wouldn't object to changing it